File: 117.populating-next-right-pointers-in-each-node-ii.cpp
/*
 * @lc app=leetcode id=117 lang=cpp
 *
 * [117] Populating Next Right Pointers in Each Node II
 */
#include <bits/stdc++.h>
using namespace std;

class Node
{
public:
    int val;
    Node *left;
    Node *right;
    Node *next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node *_left, Node *_right, Node *_next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
// @lc code=start
template <typename T>
inline T &queue_pop(queue<T> &q)
{
    T &rval = q.front();
    q.pop();
    return rval;
}

class Solution
{
public:
    Node *connect(Node *root)
    {

        if (root == NULL)
        {
            return NULL;
        }

        queue<Node *> curr_queue, next_queue;
        vector<Node *> level;
        vector<vector<Node *>> levels;

        curr_queue.push(root);
        level.push_back(root);
        levels.push_back(level);
        level = {};

        while (!curr_queue.empty())
        {
            Node *front = queue_pop(curr_queue);
            if (front->left != NULL)
            {
                level.push_back(front->left);
                next_queue.push(front->left);
            }
            if (front->right != NULL)
            {
                level.push_back(front->right);
                next_queue.push(front->right);
            }

            if (curr_queue.empty())
            {
                levels.push_back(level);
                level = {};
                swap(curr_queue, next_queue);
            }
        }

        for (auto &l : levels)
        {
            for (int i = 0; i + 1 < size(l); ++i)
            {
                l[i]->next = l[i + 1];
            }
        }

        return root;
    }
};
// @lc code=end

File: 117.populating-next-right-pointers-in-each-node-ii.py
#
# @lc app=leetcode id=117 lang=python3
#
# [117] Populating Next Right Pointers in Each Node II
#
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""


class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if root is None:
            return None
        
        curr_queue, next_queue, level, levels = ([], [], [], [])
        curr_queue.append(root)
        level.append(root)
        levels.append(level)
        level = []
        
        while len(curr_queue):
            front: Node = curr_queue.pop(0)
            
            if front.left is not None:
                level.append(front.left)
                next_queue.append(front.left)
            if front.right is not None:
                level.append(front.right)
                next_queue.append(front.right)
                
            if len(curr_queue) == 0:
                levels.append(level)
                level = []
                curr_queue = next_queue
                next_queue = []
        
        for l in levels:
            for i in range(len(l) - 1):
                l[i].next = l[i + 1]
        
        return root
        
# @lc code=end

